ledger/participant-state/kvutils/src/test/suite/scala/com/daml/ledger/participant/state/kvutils/api/BatchingLedgerWriterSpec.scala

```diff
@@ -0,0 +1,159 @@
+        res2 <- batchingWriter.commit(corId2, subm2)
+
+      } yield {
+        def verifyCommit(expected: Array[Byte]) = {
+          val commitTimeout = batchingWriter.maxWaitDuration.toMillis * 5
+          verify(mockWriter, timeout(commitTimeout))
+            .commit(ArgumentMatchers.anyString(), ArgumentMatchers.eq(expected))
+        }
+        verifyCommit(createExpectedBatch(corId1, subm1))
+        verifyCommit(createExpectedBatch(corId2, subm2))
+
+        res1 should be(SubmissionResult.Acknowledged)
+        res2 should be(SubmissionResult.Acknowledged)
+      }
+    }
+
+    "commit batch when maxBatchSizeBytes reached" in {
+      val batchCaptor: ArgumentCaptor[Array[Byte]] = ArgumentCaptor.forClass(classOf[Array[Byte]])
+      val mockWriter = createWriter(Some(batchCaptor))
+      val batchingWriter = createBatchingWriter(mockWriter, 10, 1.seconds)
+      val expectedBatch =
+        Envelope
+          .enclose(
+            DamlSubmissionBatch.newBuilder
+              .addSubmissions(DamlSubmissionBatch.CorrelatedSubmission.newBuilder
+                .setCorrelationId("test1")
+                .setSubmission(ByteString.copyFrom(Array[Byte](1, 2, 3, 4, 5))))
+              .addSubmissions(DamlSubmissionBatch.CorrelatedSubmission.newBuilder
+                .setCorrelationId("test2")
+                .setSubmission(ByteString.copyFrom(Array[Byte](6, 7, 8, 9, 10))))
+              .build)
+          .toByteArray
+
+      for {
+        res1 <- batchingWriter.commit("test1", Array[Byte](1, 2, 3, 4, 5))
+        res2 <- batchingWriter.commit("test2", Array[Byte](6, 7, 8, 9, 10))
+        res3 <- batchingWriter.commit("test3", Array[Byte](11, 12, 13, 14, 15))
+      } yield {
+        // We're only expecting the first batch to have been emitted.
+        verify(mockWriter, times(1))
+          .commit(ArgumentMatchers.anyString(), ArgumentMatchers.eq(expectedBatch))
+
+        all(Seq(res1, res2, res3)) should be(SubmissionResult.Acknowledged)
+      }
+    }
+  }
+
+  // Test that the batching writer stays up and keeps going even when underlying writer fails.
+  "drop batch when commit fails" in {
+    val mockWriter = createWriter(None, SubmissionResult.Overloaded)
+    val batchingWriter = createBatchingWriter(mockWriter, 5, 1.millis)
+    for {
+      res1 <- batchingWriter.commit("test1", Array[Byte](1, 2, 3, 4, 5))
+      res2 <- batchingWriter.commit("test2", Array[Byte](6, 7, 8, 9, 10))
+      res3 <- batchingWriter.commit("test3", Array[Byte](11, 12, 13, 14, 15))
+    } yield {
+      verify(mockWriter).commit(ArgumentMatchers.eq("test1"), ArgumentMatchers.any[Array[Byte]])
+      verify(mockWriter).commit(ArgumentMatchers.eq("test2"), ArgumentMatchers.any[Array[Byte]])
+      verify(mockWriter).commit(ArgumentMatchers.eq("test3"), ArgumentMatchers.any[Array[Byte]])
+
+      all(Seq(res1, res2, res3)) should be(SubmissionResult.Acknowledged)
+    }
+  }
+}
```

Comment:
Same as https://github.com/digital-asset/daml/pull/4964 I think, please refer to comments there